### PR TITLE
Denon Prime 4 Mapping [DISCONTINUED]

### PR DESCRIPTION
**Pull request by ultra-vi-olet**
This pull request adds 2 files: "Denon Prime 4.midi.xml" and "Denon-Prime-4-scripts.js" to the main Mixxx branch for mapping the Denon Prime 4 unit to work with Mixxx when booted into Computer Mode.

### CODE REVIEW

**Comment by uklotzde** [comment]
Thanks for publishing an early PR. Marked as *Draft* since this is considered work in progress.

Please consider rebasing on 2.3 and then force push this branch. New mappings can be merged into the most recent release and published with the next minor release.

File: res/controllers/Denon-Prime-4-scripts.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 var Prime4 = {};
+function Prime4() {}
```
Comment (Swiftb0y) [inline res/controllers/Denon-Prime-4-scripts.js]: These two lines are equivalent. A function is pretty much just an object.
```suggestion
var Prime4 = {};
```

File: res/controllers/Denon-Prime-4-scripts.js [hunk 2]
```diff
@@ -24,14 +25,126 @@ var primeHotcueColors = new ColorMapper({
...
+// code for the shift buttons
+Prime4.shiftButton = function (channel, control, value, status, group) {
+  if (value === 127) {
+    engine.connectControl(group, key, true);
+    Prime4.activeButtons = Prime4.shiftedButtons;
+    engine.connectControl(group, key);
+  } else {
+    engine.connectControl(group, key, true);
+    Prime4.activeButtons = Prime4.unshiftedButtons;
+    engine.connectControl(group, key);
```
Comment (Swiftb0y) [inline res/controllers/Denon-Prime-4-scripts.js:66]: connectControl is deprecated. prefer makeConnection instead.

File: res/controllers/Denon-Prime-4-scripts.js [hunk 3]
```diff
@@ -24,14 +25,126 @@ var primeHotcueColors = new ColorMapper({
...
+Prime4.buttons['[Channel4]'] = Prime4.buttons['[Channel2]']
+
+Prime4.channelRegEx = /\[Channel(\d+)\]/
+Prime4.deckToggleButton = function (channel, control, value, status, group) {
+  if (value) {
+    var deckNumber = parseInt(
+      Prime4.channelRegEx.exec(
+        Prime4.deck[group]
+      )[1]
+    )
```
Comment (Swiftb0y) [inline res/controllers/Denon-Prime-4-scripts.js:105]: we already provide a utility function for this: `script.deckFromGroup(group)`. See: https://github.com/mixxxdj/mixxx/wiki/midi%20scripting#Helper-functions

**Review by Swiftb0y — COMMENTED** [review]
Hey there. I know you just started your mapping so let me give you a few pointers.

Make sure you made yourself you made yourself familiar with all the resources we have on Mappings. https://github.com/mixxxdj/mixxx/wiki#controller-mapping-documentation

I highly recommend you use ComponentsJS for your mappings as it makes them much easier to review and maintain. If you want to take inspiration from other scripts, I highly recommend you look at the Roland DJ-505, A&H Xone-K2 and Behringer CMD-MM1 scripts we already have.

Thanks

**Comment by ronso0** [comment]
@whanake-music Please issue `git rebase -i` onto mixxx/2.3 in order to get rid of the git history and the merge conflicts.

**Comment by Holzhaus** [comment]
Why did you close this?

**Comment by whanake-music** [comment]
> 
> 
> Why did you close this?

I'm going to try making a new pull request that is based off the 2.3 branch to begin with. I'm having a lot of trouble with the "github rebase -i" command and I can't seem to remove the merge conflict + commit history issues. I've tried researching and figuring out how to fix it, but I think it would just be easier to restart the PR process so this one isn't cluttering everything up

**Comment by Holzhaus** [comment]
Closing the PR doesn't make a difference. A PR is just a representation of a branch. Once you have a clean branch, you could just force push and update this PR.

But closing this PR will clutter the discussion acrossmulziple PRs, and sometimes GitHub won't let you update/reopen a PR after closing it.

**Comment by whanake-music** [comment]
I didn't mean to close it at first, I did a force push after I thought I managed to remove the previous commits, but then the force push made the PR close and now i can't re-open it. it just says "The whanake-music:prime4_mapping branch has no history in common with mixxxdj:2.3." what should I do?

**Pull request closed without merging**